modal/functions.py

```diff
@@ -271,37 +312,26 @@ async def poll_function(self, timeout: Optional[float] = None):
         return await _process_result(items[0].result, items[0].data_format, self.stub, self.client)
 
     async def run_generator(self):
-        last_entry_id = "0-0"
-        completed = False
-        try:
-            while not completed:
-                request = api_pb2.FunctionGetOutputsRequest(
-                    function_call_id=self.function_call_id,
-                    timeout=OUTPUTS_TIMEOUT,
-                    last_entry_id=last_entry_id,
-                    clear_on_success=False,  # there could be more results
-                )
-                response: api_pb2.FunctionGetOutputsResponse = await retry_transient_errors(
-                    self.stub.FunctionGetOutputs,
-                    request,
-                    attempt_timeout=OUTPUTS_TIMEOUT + ATTEMPT_TIMEOUT_GRACE_PERIOD,
-                )
-                if len(response.outputs) > 0:
-                    last_entry_id = response.last_entry_id
-                    for item in response.outputs:
-                        if item.result.gen_status == api_pb2.GenericResult.GENERATOR_STATUS_COMPLETE:
-                            completed = True
-                            break
-                        yield await _process_result(item.result, item.data_format, self.stub, self.client)
-        finally:
-            # "ack" that we have all outputs we are interested in and let backend clear results
-            request = api_pb2.FunctionGetOutputsRequest(
-                function_call_id=self.function_call_id,
-                timeout=0,
-                last_entry_id="0-0",
-                clear_on_success=True,
-            )
-            await self.stub.FunctionGetOutputs(request)
+        data_stream = _stream_function_call_data(self.client, self.function_call_id, variant="data_out")
+        data_task = asyncio.create_task(data_stream.__anext__())
+        generator_finished_task = asyncio.create_task(self.run_function())
+        items_received = 0
+        items_total: Union[int, None] = None  # populated when generator_finished_task completes
+
+        while True:
+            done, _ = await asyncio.wait([data_task, generator_finished_task], return_when=asyncio.FIRST_COMPLETED)
+            if data_task in done:
+                yield data_task.result()
+                items_received += 1
+                if items_received == items_total:
+                    break
+                data_task = asyncio.create_task(data_stream.__anext__())
```

Comment:
This feels a bit messy. Could we simplify this using [aiostream](https://aiostream.readthedocs.io/en/stable/core.html)? Something like

```python
aiostream.stream.merge(data_stream, aiostream.stream.once(generator_finished_task))
```

Reply:
Will try that!

Reply:
Thanks, looks like this also fixed a bug with asynchronous shutdown thanks to the [stream context](https://aiostream.readthedocs.io/en/stable/core.html#stream-context-manager) manager